proposals/0018-clique-service.md

```diff
@@ -0,0 +1,30 @@
+---
+simd: '0018'
+title: Clique Service
+authors:
+  - Maximilian Schneider (Mango)
+category: Networking
+type: Optional
+status: Draft
+created: 2023-01-13
+---
+
+## Problem
+
+Various users don’t like the probabilistic nature of turbine, which can lead to two issues: 
+
+1. Generally high latency for RPC nodes in Eastern Asia (where most popular crypto currency exchanges are hosted)
+1. Spikes in `retransmit-stage-slot-stats.elapsed_millis` for RPC nodes, which causes delays on processing of following slots.
+
+A common workaround is to run a high stakes validator with a patch that forwards shreds directly to a selection of unstaked RPCs, in addition to the regular turbine path. This method is called “shredstream forward” and is used by Mango, Triton & Jito in production. Enabling forwards from a single staked validator in L1 of turbine can reduce latency by multiple seconds. It’s an incredibly effective method to improve an unstaked RPC node’s latency with the result is that it behaves nearly as good as if physically located in Europe.
+
+The issue with this solution is that it’s currently cumbersome to configure and hence only star shaped deployments are known to me. In addition it is very inaccessible, as it is unknown to most users and requires direct access to a high stakes validator to deploy a patched version of the solana client.
+
+## Solution
+
+Create an opt-in clique service, which in parallel to retransmit stage shares every shred received with a clique of validators. To allow for easy configuration of different network topologies the clique service can choose between different way to discover other clique members:
+
+1. public coordination servers (similar to ipfs), mango would run such a server to improve distribution of shreds in south-east Asia
+1. private lists of ips, service providers like triton could continue to run a private high performance version
+
+I would prefer to use an existing gossip network implementation, that has been battle tested over hand-rolling a third p2p network (gossip, turbine, …) and use libp2p’s [gossipsub](https://github.com/libp2p/specs/tree/master/pubsub/gossipsub). In practice I expect many users to roll their own optimized implementation, but at least now they have a central point they can just override, rather than having to manually patch diferent points in the code.
```

Comment:
Bandwitdth & fanout considerations. I can run this service in a star topology for maybe 10 nodes. The load from each connected node is way higher than from a turbine neighbour/child, as we are now forwarding everything rather than a subset. At some point we'll get bandwidth limited.

The geographical proximity configuration allows for the assumption that latency between nodes is below 50ms, hence even submission across 4 hops can still be bound by 200ms latency. The latency lag we are trying to battle is in seconds no 100ms for reference.

We considered continuing the current approach of "just adding retransmit ips" but that's already done and running in production. I think the part that is interesting here is to work out a path where different parties can quickly experiment with different protocols to improve local distribution, rather than causing tight coupling with the existing code base.

Reply:
Sorry, this is still not clear to me how it would operate in practice.

Lets say I am running a high staked node and I am ok to forward shreds I receive from turbine to a few other nodes. In this protocol you are proposing
1. What type of messages do I receive from those other nodes who want me to forward shreds to them?
2. What other type of messages (aside from the above and the shreds) are exchanged through this protocol?
3. How would I decide of the nodes who want me to forward shreds to them which ones to acknowledge and which ones don't?

Reply:
>1. heartbeats to know if they are online or not

This is currently sort of done with periodic `ContactInfo` refresh in gossip:
https://github.com/solana-labs/solana/blob/b35d173c0/gossip/src/cluster_info.rs#L1764-L1765
https://github.com/solana-labs/solana/blob/b35d173c0/gossip/src/cluster_info.rs#L443-L457

There are also ping/pong messages sent between nodes periodically:
https://github.com/solana-labs/solana/blob/b35d173c0/gossip/src/ping_pong.rs#L179-L204

For stake nodes we can also look up their last voted slot from bank as a signal if the node is up and running or not:
https://github.com/solana-labs/solana/blob/b35d173c0/core/src/consensus.rs#L450-L454

>2. network topology metrics, like inter-node latency, decent optimization needs to be topology aware

We do send ping/pong messages between nodes in gossip; and track when the ping was sent out:
https://github.com/solana-labs/solana/blob/b35d173c0/gossip/src/ping_pong.rs#L43-L45
Potentially we can utilize this ping/pong round-trip time as a proxy for inter-node latency:
https://github.com/solana-labs/solana/blob/b35d173c0/gossip/src/ping_pong.rs#L162-L177

Another proxy could be just based on shreds delivery time themselves.

Reply:
I agree that these existing methods could be used, thank you for proposing those. I can see how we leverage them, once we conducted enough experiments to settle on a reduced feature set.

For now we would like to experiment very quickly with a stable interface towards solana core, during development of this service. In order to minimize impact on existing validator functionality, I would prefer to decouple concerns as much as possible.

- Solana gossip is a core feature to maximize cluster availability & decentralization.
- Clique is an optional feature to improve latency and can be executed by bankless nodes.

Here's an example branch that explores how such an interface could look like:

https://github.com/blockworks-foundation/solana/pull/19